### PR TITLE
374 log redaction

### PR DESCRIPTION
**Pull request by sqweelygig**
* SyncBot: Log the full error response from services.
* ProcBot: Add function to calculate the key/value pairs that would be injected.
* ServiceScaffold: Store logger instance and construction data when constructed.
* ServiceScaffold: Bubble to error handler details of the request and the session.
* Logger: Allow configuration of a list of strings to redact.
* Logger: Move log redaction to the underlying, universal, function to catch more outputs.

Test routine:
--

1) For each of Discourse, Front & Flowdock:
    1) Introduce a deliberate error in the payload delivered to the service
    1) Capture an output of that error occurring on the master branch
    1) Capture an output of that error occurring on the feature branch
    1) Confirm that I can see the error in the second capture
    1) Confirm that secrets are not present in the second capture
1) Remove the captures via their own commit, so they can be cited for now, but disappear on merge

Change-Type: patch
Connects-To: #374
See: https://github.com/resin-io-modules/resin-procbots/commit/a915cbc983d089e3d7ddcabb9ff890859cc37e86

---
##### Contributor checklist

- [x] Build output been rebuilt and tested
- [ ] Introduces security considerations
- [x] Tests are included
- [ ] Documentation is added or changed
- [ ] Affects the development, build or deployment processes of the component

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---
---- Autogenerated Waffleboard Connection: Connects to #374

### CODE REVIEW

**Comment by sqweelygig** [comment]
The secrets are controlled by the configuration of the server, not by the people who interact with SyncBot.

**Comment by lurch** [comment]
Pinging @izavits as this might tangentially connect to https://github.com/resin-io/hq/pull/897 ?

File: lib/framework/procbot.ts [hunk 1]
```diff
@@ -58,29 +59,52 @@ export class ProcBot {
 			return _.mapValues(value, ProcBot.injectEnvironmentVariables);
 		}
 		if (_.isString(value)) {
-			return ProcBot.stringInjector(value);
+			return ProcBot.stringInjector(value).cooked;
 		}
 		return value;
 	}
 
+	public static calculateInjections(value: any): Dictionary<string> {
```
Comment (hedss) [inline lib/framework/procbot.ts]: Don't like `calculateInjections` as a name as no actual transformation is occurring. `determineInjections`, perhaps?

File: lib/utils/logger.ts [hunk 2]
```diff
@@ -132,6 +134,12 @@ export class Logger {
 	// Generic output method for either type.
 	private output(level: number, classLevel: number, levelStrings: string[], message: string): void {
 		if (level <= classLevel) {
+			if (!_.isEmpty(this._secrets) && process.env.PROCBOT_LOG_SECRETS !== 'true') {
```
Comment (hedss) [inline lib/utils/logger.ts]: Were we not going to base this on the output level? I don't really like having a new envvar specifically for logging secrets.

File: lib/utils/logger.ts [hunk 3]
```diff
@@ -58,6 +58,7 @@ export class Logger {
 	// These can be overriden by specific methods.
 	private _logLevel = _.parseInt(process.env.PROCBOT_LOG_LEVEL) || LogLevel.INFO;
 	private _alertLevel = _.parseInt(process.env.PROCBOT_ALERT_LEVEL) || AlertLevel.ERROR;
+	private _secrets: string[] = [];
```
Comment (hedss) [inline lib/utils/logger.ts]: This doesn't need to be `_secrets` as there's no setter/getter with the same name.

File: lib/utils/logger.ts [hunk 4]
```diff
@@ -104,19 +105,20 @@ export class Logger {
 		this._alertLevel = level;
 	}
 
+	/**
+	 * Add a list of strings to the list of strings to redact.
+	 * @param secrets  Secrets to redact from any output.
+	 */
+	public pushSecrets(secrets: string[]): void {
```
Comment (hedss) [inline lib/utils/logger.ts]: However, considering the comment above, arguably this should be a setter:
```typescript
public set secrets(...) { ... }
```
In which case, the class member name *should* be `_secrets`. :)

File: lib/utils/logger.ts (hunk 3 above)
Comment (sqweelygig) [reply lib/utils/logger.ts]: Note to self: ignored as per next comment.

**Pull request merged**